File: modules/biovision-facephys/ios/HybridFacePhysOutput.swift
import AVFoundation
import CoreImage
import NitroModules
import VisionCamera

private final class SampleBufferDelegate: NSObject, AVCaptureVideoDataOutputSampleBufferDelegate {
  private let onSampleBuffer: (CMSampleBuffer, AVCaptureConnection) -> Void

  init(onSampleBuffer: @escaping (CMSampleBuffer, AVCaptureConnection) -> Void) {
    self.onSampleBuffer = onSampleBuffer
  }

  func captureOutput(_ output: AVCaptureOutput, didOutput sampleBuffer: CMSampleBuffer, from connection: AVCaptureConnection) {
    onSampleBuffer(sampleBuffer, connection)
  }
}

/// A VisionCamera output that runs the on-device face pipeline: MediaPipe Face Landmarker
/// for the live mesh and, when `measureVitals` is set, FacePhys heart rate. Camera frames
/// never leave this class; only landmarks, BVP samples and estimates reach JS.
final class HybridFacePhysOutput: HybridCameraOutputSpec, NativeCameraOutput {
  private static let workingWidth = 480
  private static let frameIntervalMs = 1000.0 / 30.0
  private static let updateIntervalMs = 66.0

  let output = AVCaptureVideoDataOutput()
  let requiresAudioInput = false
  let requiresDepthFormat = false
  let mediaType: MediaType = .video
  let streamType: StreamType = .video
  var outputOrientation: CameraOrientation = .up
  var currentResolution: Size? {
    output.connection(with: .video)?.inputStreamResolution
  }
  var targetResolution: ResolutionRule { .closestTo(Size(width: 720, height: 1280)) }

  private let queue = DispatchQueue(label: "biovision.facepipeline", qos: .userInitiated)
  private let measureVitals: Bool
  private let onUpdate: (FacePhysUpdate) -> Void
  private let onError: (Error) -> Void
  private var delegate: SampleBufferDelegate?

  // No color management: models must see the camera's own pixel values, as the browser canvas does.
  private let ciContext = CIContext(options: [.cacheIntermediates: false, .workingColorSpace: NSNull(), .outputColorSpace: NSNull()])
  private var pool: CVPixelBufferPool?
  private var poolSize = (width: 0, height: 0)
  private var connectionConfigured = false

  private var detector: BlazeFaceDetector?
  private var landmarker: FaceLandmarker?
  private var engine: FacePhysEngine?
  private var failed = false

  private var lastFrameMs = 0.0
  private var lastUpdateMs = 0.0
  private var frameIndex = 0
  private var frameTimes: [Double] = []
  private var pending: [FacePhysSample] = []
  private var latestLandmarks: [Double]?
  private var latestFace: FacePhysFaceBox?
  private var latestHeartRate: Double?
  private var latestQuality: Double?

  init(options: FacePhysOutputOptions) {
    measureVitals = options.measureVitals
    onUpdate = options.onUpdate
    onError = options.onError
    super.init()
    output.videoSettings = [kCVPixelBufferPixelFormatTypeKey as String: kCVPixelFormatType_32BGRA]
    output.alwaysDiscardsLateVideoFrames = true
    let delegate = SampleBufferDelegate { [weak self] buffer, connection in self?.handle(buffer, connection) }
    self.delegate = delegate
    output.setSampleBufferDelegate(delegate, queue: queue)
  }

  func configure(config: OutputConfiguration) {
    guard let connection = output.connection(with: .video) else { return }
    configureConnection(connection)
  }

  /// Upright, unmirrored frames. Also applied on the first delivered frame, because the
  /// connection may not exist yet when VisionCamera calls `configure`.
  private func configureConnection(_ connection: AVCaptureConnection) {
    connection.preferredVideoStabilizationMode = .off
    if connection.isVideoOrientationSupported { connection.videoOrientation = .portrait }
    if connection.isVideoMirroringSupported {
      connection.automaticallyAdjustsVideoMirroring = false
      connection.isVideoMirrored = false
    }
    connectionConfigured = true
  }

  private func handle(_ buffer: CMSampleBuffer, _ connection: AVCaptureConnection) {
    if failed { return }
    if !connectionConfigured { configureConnection(connection) }
    let captureMs = CMTimeGetSeconds(CMSampleBufferGetPresentationTimeStamp(buffer)) * 1000
    // 30 fps, like the web frame loop; tolerate a millisecond of camera jitter.
    if lastFrameMs > 0 && captureMs - lastFrameMs < HybridFacePhysOutput.frameIntervalMs - 1 { return }
    lastFrameMs = captureMs

    do {
      if detector == nil {
        detector = try BlazeFaceDetector()
        landmarker = try FaceLandmarker()
        if measureVitals { engine = try FacePhysEngine() }
      }
      guard let source = CMSampleBufferGetImageBuffer(buffer), let frame = try uprightFrame(source) else { return }
      CVPixelBufferLockBaseAddress(frame, .readOnly)
      defer { CVPixelBufferUnlockBaseAddress(frame, .readOnly) }
      guard let base = CVPixelBufferGetBaseAddress(frame) else { return }
      try process(
        pixels: base.assumingMemoryBound(to: UInt8.self),
        width: CVPixelBufferGetWidth(frame),
        height: CVPixelBufferGetHeight(frame),
        bytesPerRow: CVPixelBufferGetBytesPerRow(frame),
        captureMs: captureMs
      )
    } catch {
      failed = true
      onError(error)
    }
  }

  private func process(pixels: UnsafePointer<UInt8>, width: Int, height: Int, bytesPerRow: Int, captureMs: Double) throws {
    guard let detector, let landmarker else { return }
    frameIndex += 1
    frameTimes.append(captureMs)
    frameTimes.removeAll { captureMs - $0 > 1000 }

    // FacePhys needs a BlazeFace box every frame (as on web); share it with the landmarker.
    var detection: FaceDetection?
    var detected = false
    let detect = { () throws -> FaceDetection? in
      if !detected { detection = try detector.detect(bgra: pixels, width: width, height: height, bytesPerRow: bytesPerRow); detected = true }
      return detection
    }

    // With vitals running, the mesh updates at 15 fps so heart-rate frames stay on time.
    if !measureVitals || frameIndex % 2 == 0 {
      if let points = try landmarker.process(bgra: pixels, width: width, height: height, bytesPerRow: bytesPerRow, detection: detect) {
        var normalized = [Double](repeating: 0, count: points.count)
        for index in 0..<478 {
          normalized[index * 3] = Double(points[index * 3]) / Double(width)
          normalized[index * 3 + 1] = Double(points[index * 3 + 1]) / Double(height)
          normalized[index * 3 + 2] = Double(points[index * 3 + 2]) / Double(width)
        }
        latestLandmarks = normalized
      } else {
        latestLandmarks = nil
      }
    }

    if let engine {
      let face = try detect()?.rect
      let result = try engine.process(bgra: pixels, width: width, height: height, bytesPerRow: bytesPerRow, captureMs: captureMs, face: face)
      if let sample = result.sample { pending.append(FacePhysSample(value: Double(sample.value), timestampMs: sample.timestampMs)) }
      if let heartRate = result.heartRate { latestHeartRate = Double(heartRate) }
      if let quality = result.signalQuality { latestQuality = Double(quality) }
      latestFace = face.map {
        FacePhysFaceBox(x: Double($0.x) / Double(width), y: Double($0.y) / Double(height),
                        width: Double($0.width) / Double(width), height: Double($0.height) / Double(height))
      }
    }

    guard captureMs - lastUpdateMs >= HybridFacePhysOutput.updateIntervalMs else { return }
    lastUpdateMs = captureMs
    let update = FacePhysUpdate(
      frameWidth: Double(width),
      frameHeight: Double(height),
      landmarks: latestLandmarks,
      samples: pending,
      heartRate: latestHeartRate,
      signalQuality: latestQuality,
      face: latestFace,
      framesPerSecond: Double(frameTimes.count)
    )
    // Estimates are sent once per PSD run so JS sees one reading per spectrum, as on web.
    pending.removeAll(keepingCapacity: true)
    latestHeartRate = nil
    latestQuality = nil
    onUpdate(update)
  }

  /// Any camera pixel format and orientation -> upright BGRA, 480 px wide. Rotates sensor-
  /// oriented (landscape) buffers in case the portrait connection setting was not honoured.
  private func uprightFrame(_ source: CVPixelBuffer) throws -> CVPixelBuffer? {
    var image = CIImage(cvPixelBuffer: source)
    if image.extent.width > image.extent.height { image = image.oriented(.right) }
    let scale = CGFloat(HybridFacePhysOutput.workingWidth) / image.extent.width
    image = image.transformed(by: CGAffineTransform(scaleX: scale, y: scale))
    image = image.transformed(by: CGAffineTransform(translationX: -image.extent.origin.x, y: -image.extent.origin.y))
    let width = HybridFacePhysOutput.workingWidth
    let height = Int((image.extent.height).rounded())
    guard height > 0 else { return nil }

    if pool == nil || poolSize.width != width || poolSize.height != height {
      let attributes: [String: Any] = [
        kCVPixelBufferPixelFormatTypeKey as String: kCVPixelFormatType_32BGRA,
        kCVPixelBufferWidthKey as String: width,
        kCVPixelBufferHeightKey as String: height,
        kCVPixelBufferIOSurfacePropertiesKey as String: [:],
      ]
      var created: CVPixelBufferPool?
      CVPixelBufferPoolCreate(nil, nil, attributes as CFDictionary, &created)
      pool = created
      poolSize = (width, height)
    }
    guard let pool else { return nil }
    var target: CVPixelBuffer?
    CVPixelBufferPoolCreatePixelBuffer(nil, pool, &target)
    guard let target else { return nil }
    ciContext.render(image, to: target, bounds: CGRect(x: 0, y: 0, width: width, height: height), colorSpace: nil)
    return target
  }
}

File: modules/biovision-facephys/android/src/main/java/com/biovision/facephys/HybridFacePhysOutput.kt
package com.biovision.facephys

import android.content.Context
import android.hardware.camera2.CameraCharacteristics
import android.hardware.camera2.CameraManager
import android.hardware.camera2.CaptureRequest
import android.util.Log
import android.util.Size as AndroidSize
import androidx.annotation.OptIn
import androidx.camera.camera2.interop.Camera2Interop
import androidx.camera.camera2.interop.ExperimentalCamera2Interop
import androidx.camera.core.ImageAnalysis
import androidx.camera.core.ImageProxy
import androidx.camera.core.resolutionselector.ResolutionSelector
import androidx.camera.core.resolutionselector.ResolutionStrategy
import com.margelo.nitro.biovision.facephys.FacePhysFaceBox
import com.margelo.nitro.biovision.facephys.FacePhysOutputOptions
import com.margelo.nitro.biovision.facephys.FacePhysSample
import com.margelo.nitro.biovision.facephys.FacePhysUpdate
import com.margelo.nitro.camera.CameraOrientation
import com.margelo.nitro.camera.HybridCameraOutputSpec
import com.margelo.nitro.camera.MediaType
import com.margelo.nitro.camera.MirrorMode
import com.margelo.nitro.camera.Size
import com.margelo.nitro.camera.public.NativeCameraOutput
import java.util.concurrent.Executors
import kotlin.math.max
import kotlin.math.roundToInt

/**
 * A VisionCamera output that runs the on-device face pipeline: MediaPipe Face Landmarker for
 * the live mesh and, when `measureVitals` is set, FacePhys heart rate. Port of
 * ios/HybridFacePhysOutput.swift.
 *
 * Camera frames never leave this class; only landmarks, BVP samples and estimates reach JS.
 *
 * Where iOS drives an AVCaptureVideoDataOutput and rotates frames itself with CoreImage, this
 * uses a CameraX ImageAnalysis with `setOutputImageRotationEnabled`, so the buffers arrive
 * upright, and downscales them to the same 480 px working width.
 */
class HybridFacePhysOutput(
  private val context: Context,
  private val options: FacePhysOutputOptions,
) : HybridCameraOutputSpec(),
  NativeCameraOutput {

  private companion object {
    const val TAG = "BioVisionFacePhys"
    const val WORKING_WIDTH = 480
    const val FRAME_INTERVAL_MS = 1000.0 / 30.0
    const val UPDATE_INTERVAL_MS = 66.0
  }

  private val executor = Executors.newSingleThreadExecutor { runnable ->
    Thread(runnable, "biovision.facepipeline")
  }

  override val mediaType: MediaType = MediaType.VIDEO
  override var outputOrientation: CameraOrientation = CameraOrientation.UP
  override val currentResolution: Size?
    get() = imageAnalysis?.resolutionInfo?.resolution?.let { Size(it.width.toDouble(), it.height.toDouble()) }
  override var mirrorMode: MirrorMode = MirrorMode.AUTO

  private var imageAnalysis: ImageAnalysis? = null

  private var detector: BlazeFaceDetector? = null
  private var landmarker: FaceLandmarker? = null
  private var engine: FacePhysEngine? = null
  private var failed = false

  private var lastFrameMs = 0.0
  private var lastUpdateMs = 0.0
  private var frameIndex = 0
  private val frameTimes = ArrayList<Double>()
  private val pending = ArrayList<FacePhysSample>()
  private var latestLandmarks: DoubleArray? = null
  private var latestFace: FacePhysFaceBox? = null
  private var latestHeartRate: Double? = null
  private var latestQuality: Double? = null

  /** Reused scratch buffers, so a 30 fps pipeline does not allocate per frame. */
  private var working = ByteArray(0)
  private var workingWidth = 0
  private var workingHeight = 0
  private var source = ByteArray(0)
  private var columnOffsets: IntArray? = null
  private var columnSourceWidth = 0

  /** Rolling diagnostics, logged so a release build on a real phone can be diagnosed. */
  private var diagnosticsAtMs = 0.0
  private var framesSinceDiagnostics = 0

  @OptIn(ExperimentalCamera2Interop::class)
  override fun createUseCase(
    mirrorMode: MirrorMode,
    config: NativeCameraOutput.Config,
  ): NativeCameraOutput.PreparedUseCase {
    // iOS asks for 720x1280 and downsizes each frame on the GPU with CoreImage. Android has no
    // equivalent here, so ask the camera for a stream near the 480 px working width instead:
    // scaling 720x1280 in Kotlin on every frame costs more than the pipeline can afford at 30 fps,
    // and a slow pipeline starves the rPPG model of the frame rate it was trained on.
    val resolutionSelector = ResolutionSelector.Builder()
      .setResolutionStrategy(
        ResolutionStrategy(
          AndroidSize(WORKING_WIDTH, WORKING_WIDTH * 4 / 3),
          ResolutionStrategy.FALLBACK_RULE_CLOSEST_HIGHER_THEN_LOWER,
        ),
      )
      .build()

    val builder = ImageAnalysis.Builder()
      .setResolutionSelector(resolutionSelector)
      .setBackpressureStrategy(ImageAnalysis.STRATEGY_KEEP_ONLY_LATEST)
      .setOutputImageFormat(ImageAnalysis.OUTPUT_IMAGE_FORMAT_RGBA_8888)
      // Upright, unmirrored buffers, matching the portrait connection iOS configures.
      .setOutputImageRotationEnabled(true)
      .setBackgroundExecutor(executor)

    // Mains flicker at 50 Hz beats against a 30 fps stream and produces a slow brightness
    // oscillation that both looks like a pulse and drowns the real one, so antibanding is pinned
    // rather than left on auto. The frame rate is pinned to 30 because that is where the model
    // was trained, and a rate that sags in dim light changes what the model is fed.
    //
    // White balance is deliberately NOT locked. Locking it freezes whatever tint the camera
    // happened to hold when the session opened, which tinted the whole preview green — these
    // options apply to every stream in the session, not only to this analysis output.
    val extender = Camera2Interop.Extender(builder)
      .setCaptureRequestOption(
        CaptureRequest.CONTROL_AE_ANTIBANDING_MODE,
        CaptureRequest.CONTROL_AE_ANTIBANDING_MODE_50HZ,
      )
      .setCaptureRequestOption(CaptureRequest.CONTROL_AE_TARGET_FPS_RANGE, android.util.Range(30, 30))
      // iOS turns stabilisation off for the same reason: warping the frame moves the skin
      // region between frames, which reads as signal.
      .setCaptureRequestOption(
        CaptureRequest.CONTROL_VIDEO_STABILIZATION_MODE,
        CaptureRequest.CONTROL_VIDEO_STABILIZATION_MODE_OFF,
      )

    // The pulse shows up as roughly a quarter of one 8-bit level of brightness change. Android's
    // image pipeline is tuned to make video look good, and two of its stages are larger than that
    // signal: temporal noise reduction averages successive frames, which attenuates exactly the
    // high-frequency variation the model reads, and edge enhancement sharpens what is left.
    // Temporal denoising also ramps up over the first seconds, which is why a scan can start
    // usable and then decay. Both are turned off where the device allows it.
    noiseReductionMode()?.let { extender.setCaptureRequestOption(CaptureRequest.NOISE_REDUCTION_MODE, it) }
    edgeMode()?.let { extender.setCaptureRequestOption(CaptureRequest.EDGE_MODE, it) }

    val analysis = builder.build()

    analysis.setAnalyzer(executor) { image -> handle(image) }

    return NativeCameraOutput.PreparedUseCase(analysis) {
      this.imageAnalysis = analysis
      this.mirrorMode = mirrorMode
    }
  }

  /** Characteristics of the front camera, which is the only one this output is used with. */
  private fun frontCameraCharacteristics(): CameraCharacteristics? = runCatching {
    val manager = context.getSystemService(Context.CAMERA_SERVICE) as CameraManager
    val id = manager.cameraIdList.firstOrNull { cameraId ->
      manager.getCameraCharacteristics(cameraId)
        .get(CameraCharacteristics.LENS_FACING) == CameraCharacteristics.LENS_FACING_FRONT
    }
    id?.let { manager.getCameraCharacteristics(it) }
  }.getOrNull()

  /**
   * The least aggressive noise reduction the device offers. OFF is not guaranteed to exist, so
   * fall back to MINIMAL and then FAST rather than requesting a mode the camera will reject.
   */
  private fun noiseReductionMode(): Int? {
    val available = frontCameraCharacteristics()
      ?.get(CameraCharacteristics.NOISE_REDUCTION_AVAILABLE_NOISE_REDUCTION_MODES)
      ?.toSet()
      ?: return null
    val choice = listOf(
      CaptureRequest.NOISE_REDUCTION_MODE_OFF,
      CaptureRequest.NOISE_REDUCTION_MODE_MINIMAL,
      CaptureRequest.NOISE_REDUCTION_MODE_FAST,
    ).firstOrNull { it in available }
    Log.i(TAG, "noise reduction mode: ${choice ?: "device default"}")
    return choice
  }

  /** Same idea for edge enhancement: sharpening amplifies noise on top of a very small signal. */
  private fun edgeMode(): Int? {
    val available = frontCameraCharacteristics()
      ?.get(CameraCharacteristics.EDGE_AVAILABLE_EDGE_MODES)
      ?.toSet()
      ?: return null
    val choice = listOf(CaptureRequest.EDGE_MODE_OFF, CaptureRequest.EDGE_MODE_FAST)
      .firstOrNull { it in available }
    Log.i(TAG, "edge mode: ${choice ?: "device default"}")
    return choice
  }

  private fun handle(image: ImageProxy) {
    try {
      if (failed) return
      val captureMs = image.imageInfo.timestamp / 1_000_000.0
      // 30 fps, like the web frame loop; tolerate a millisecond of camera jitter.
      if (lastFrameMs > 0 && captureMs - lastFrameMs < FRAME_INTERVAL_MS - 1) return
      lastFrameMs = captureMs

      if (detector == null) {
        detector = BlazeFaceDetector(context)
        landmarker = FaceLandmarker(context)
        if (options.measureVitals) engine = FacePhysEngine(context)
      }

      downscale(image)
      process(captureMs)
    } catch (error: Throwable) {
      failed = true
      options.onError.invoke(error as? Exception ?: RuntimeException(error))
    } finally {
      image.close()
    }
  }

  /**
   * Any incoming resolution -> tightly packed RGBA at the working width.
   *
   * Two things matter for keeping 30 fps on mid-range hardware: the scratch buffers are reused
   * rather than reallocated per frame, and a frame that already arrives at the working width is
   * only row-copied (CameraX pads rows, so the stride still has to be removed).
   */
  private fun downscale(image: ImageProxy) {
    val plane = image.planes[0]
    val buffer = plane.buffer
    val sourceStride = plane.rowStride
    val sourceWidth = image.width
    val sourceHeight = image.height

    val needed = buffer.remaining()
    if (source.size < needed) source = ByteArray(needed)
    buffer.get(source, 0, needed)

    val targetWidth = minOf(WORKING_WIDTH, sourceWidth)
    val targetHeight = max(1, (sourceHeight.toLong() * targetWidth / sourceWidth).toInt())
    if (workingWidth != targetWidth || workingHeight != targetHeight) {
      workingWidth = targetWidth
      workingHeight = targetHeight
      working = ByteArray(targetWidth * targetHeight * 4)
    }

    if (targetWidth == sourceWidth && targetHeight == sourceHeight) {
      for (row in 0 until targetHeight) {
        System.arraycopy(source, row * sourceStride, working, row * targetWidth * 4, targetWidth * 4)
      }
      return
    }

    // Nearest-neighbour: the models consume a 36x36 area average and a 256x256 bilinear crop of
    // this frame, so the extra smoothing a bilinear pass would add here is not worth its cost.
    val xMap = columnMap(sourceWidth, targetWidth)
    for (row in 0 until targetHeight) {
      val sourceY = ((row * sourceHeight) / targetHeight).coerceAtMost(sourceHeight - 1)
      val sourceRow = sourceY * sourceStride
      var target = row * targetWidth * 4
      for (column in 0 until targetWidth) {
        var offset = sourceRow + xMap[column]
        working[target] = source[offset]
        working[target + 1] = source[offset + 1]
        working[target + 2] = source[offset + 2]
        working[target + 3] = source[offset + 3]
        target += 4
      }
    }
  }

  /** Source byte offsets for each output column, computed once per resolution. */
  private fun columnMap(sourceWidth: Int, targetWidth: Int): IntArray {
    val cached = columnOffsets
    if (cached != null && cached.size == targetWidth && columnSourceWidth == sourceWidth) return cached
    val map = IntArray(targetWidth) { column ->
      (((column * sourceWidth) / targetWidth).coerceAtMost(sourceWidth - 1)) * 4
    }
    columnOffsets = map
    columnSourceWidth = sourceWidth
    return map
  }

  private fun process(captureMs: Double) {
    val detector = detector ?: return
    val landmarker = landmarker ?: return
    val width = workingWidth
    val height = workingHeight
    val rowStride = width * 4

    frameIndex += 1
    frameTimes.add(captureMs)
    frameTimes.removeAll { captureMs - it > 1000 }

    // FacePhys needs a BlazeFace box every frame (as on web); share it with the landmarker.
    var detection: FaceDetection? = null
    var detected = false
    val detect = {
      if (!detected) {
        detection = detector.detect(working, width, height, rowStride)
        detected = true
      }
      detection
    }

    // With vitals running, the mesh updates at 15 fps so heart-rate frames stay on time.
    if (!options.measureVitals || frameIndex % 2 == 0) {
      val points = landmarker.process(working, width, height, rowStride, detect)
      latestLandmarks = if (points != null) {
        DoubleArray(points.size).also { normalized ->
          for (index in 0 until 478) {
            normalized[index * 3] = points[index * 3].toDouble() / width
            normalized[index * 3 + 1] = points[index * 3 + 1].toDouble() / height
            normalized[index * 3 + 2] = points[index * 3 + 2].toDouble() / width
          }
        }
      } else {
        null
      }
    }

    val engine = engine
    if (engine != null) {
      val face = detect()?.rect
      val result = engine.process(working, width, height, rowStride, captureMs, face)
      result.sample?.let { pending.add(FacePhysSample(it.value.toDouble(), it.timestampMs)) }
      result.heartRate?.let { latestHeartRate = it.toDouble() }
      result.signalQuality?.let { latestQuality = it.toDouble() }
      latestFace = face?.let {
        FacePhysFaceBox(
          it.x.toDouble() / width,
          it.y.toDouble() / height,
          it.width.toDouble() / width,
          it.height.toDouble() / height,
        )
      }
    }

    framesSinceDiagnostics += 1
    // One line every 3 s, in release too: without a device attached this is the only way to see
    // whether the pipeline is keeping frame rate and what the signal-quality model reports.
    if (captureMs - diagnosticsAtMs >= 3000) {
      val seconds = if (diagnosticsAtMs > 0) (captureMs - diagnosticsAtMs) / 1000.0 else 3.0
      Log.i(
        TAG,
        "${width}x${height} @ ${"%.1f".format(framesSinceDiagnostics / seconds)} fps" +
          ", face=${if (latestFace != null) "yes" else "no"}" +
          ", mesh=${if (latestLandmarks != null) "yes" else "no"}" +
          ", dt=${engine?.let { "%.1fms".format(it.modelFrameIntervalMs) } ?: "-"}" +
          ", green=${engine?.let { "%.4f".format(it.lastCropGreen) } ?: "-"}" +
          ", sqi=${latestQuality?.let { "%.2f".format(it) } ?: "-"}" +
          ", hr=${latestHeartRate?.let { "%.1f".format(it) } ?: "-"}",
      )
      diagnosticsAtMs = captureMs
      framesSinceDiagnostics = 0
    }

    if (captureMs - lastUpdateMs < UPDATE_INTERVAL_MS) return
    lastUpdateMs = captureMs
    val update = FacePhysUpdate(
      width.toDouble(),
      height.toDouble(),
      latestLandmarks,
      pending.toTypedArray(),
      latestHeartRate,
      latestQuality,
      latestFace,
      frameTimes.size.toDouble(),
    )
    // Estimates are sent once per PSD run so JS sees one reading per spectrum, as on web.
    pending.clear()
    latestHeartRate = null
    latestQuality = null
    options.onUpdate.invoke(update)
  }

  override fun dispose() {
    super.dispose()
    imageAnalysis?.clearAnalyzer()
    detector?.close()
    landmarker?.close()
    engine?.close()
    executor.shutdown()
  }
}
